server/sonar-server/src/main/java/org/sonar/server/rule/index/RuleIndexer.java

```diff
@@ -56,84 +68,175 @@ public RuleIndexer(EsClient esClient, DbClient dbClient) {
-    if (uninitializedIndexTypes.contains(INDEX_TYPE_RULE_EXTENSION)) {
-      try (RuleMetadataIterator metadatas = new RuleMetadataIterator(dbClient)) {
-        doIndexRuleExtensions(metadatas, bulk);
+      // index all organization extensions
+      if (uninitializedIndexTypes.contains(INDEX_TYPE_RULE_EXTENSION)) {
+        dbClient.ruleDao().scrollIndexingRuleExtensions(dbSession, dto -> {
+          bulk.add(newRuleExtensionDocIndexRequest(dto));
+        });
       }
+
+      bulk.stop();
     }
+  }
 
-    bulk.stop();
+  public void commitAndIndex(DbSession dbSession, RuleKey ruleKey) {
+    commitAndIndex(dbSession, singletonList(ruleKey));
   }
 
-  public void indexRuleDefinition(RuleKey ruleKey) {
-    indexRuleDefinitions(singletonList(ruleKey));
+  public void commitAndIndex(DbSession dbSession, Collection<RuleKey> ruleDtos) {
+    List<EsQueueDto> items = ruleDtos.stream()
+      .map(key -> EsQueueDto.create(EsQueueDto.Type.RULE, key.toString()))
+      .collect(MoreCollectors.toArrayList());
+
+    dbClient.esQueueDao().insert(dbSession, items);
+    dbSession.commit();
+    postCommit(dbSession, ruleDtos, items);
   }
 
-  public void indexRuleDefinitions(List<RuleKey> ruleKeys) {
-    BulkIndexer bulk = new BulkIndexer(esClient, RuleIndexDefinition.INDEX, Size.REGULAR);
-    bulk.start();
+  public void commitAndIndex(DbSession dbSession, OrganizationDto organizationDto, RuleKey ruleKey) {
+    List<EsQueueDto> items = singletonList(EsQueueDto.create(EsQueueDto.Type.RULE_EXTENSION, ruleKey + "|" + organizationDto.getUuid()));
 
-    try (RuleIterator rules = new RuleIteratorForMultipleChunks(dbClient, ruleKeys)) {
-      doIndexRuleDefinitions(rules, bulk);
-    }
+    dbClient.esQueueDao().insert(dbSession, items);
+    dbSession.commit();
+    postCommit(dbSession, ruleKey, organizationDto, items);
+  }
 
-    bulk.stop();
+  /**
+   * Entry point for Byteman tests. See directory tests/resilience.
+   * The parameter "ruleKeys" is used only by the Byteman script.
+   */
+  private void postCommit(DbSession dbSession, Collection<RuleKey> ruleKeys, Collection<EsQueueDto> items) {
+    index(dbSession, items);
   }
 
-  public void indexRuleExtension(OrganizationDto organization, RuleKey ruleKey) {
-    try (DbSession dbSession = dbClient.openSession(false)) {
-      dbClient.ruleDao()
-        .selectMetadataByKey(dbSession, ruleKey, organization)
-        .map(ruleExtension -> RuleExtensionDoc.of(ruleKey, RuleExtensionScope.organization(organization), ruleExtension))
-        .map(Arrays::asList)
-        .map(List::iterator)
-        .ifPresent(metadata -> {
-          BulkIndexer bulk = new BulkIndexer(esClient, RuleIndexDefinition.INDEX, Size.REGULAR);
-          bulk.start();
-          doIndexRuleExtensions(metadata, bulk);
-          bulk.stop();
-        });
-    }
+  private void postCommit(DbSession dbSession, RuleKey ruleKeys, OrganizationDto organizationDto, Collection<EsQueueDto> items) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "ruleKeys". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)